### ISSUE TITLE
Distribution and Funding the Development of Mir

### ISSUE THREAD

**Issue by bpfarmer** [mir-core/mips#8]
A core question for any decentralized protocol is: how to fairly distribute tokens, particularly to the group of contributors that builds the protocol. This group could include any combination of: a for-profit company that owns the trademark and develops reference implementations, as in Zcash, a foundation that distributes grants to independent developers and organizations, as in Ethereum and Zcash, or independent developers or organizations that contribute to open source projects. I'd like to examine the various models for distribution within a decentralized protocol and the competing arguments for each with respect to decentralization, fairness, and effectiveness.

## Distribution Models ##

There would appear to be three distinct models for distributing tokens in a decentralized protocol.

1. No explicit distribution of tokens to early contributors or developers, as in Bitcoin - distribution consists entirely of block rewards and developers contribute to the protocol out of generosity, interest, or in exchange for donations from early holders.

2. A pre-mine, in which a significant percentage of the early supply is distributed to a core set of contributors or a foundation, as in Ethereum, or distributed to unrelated open source projects, as in Handshake (I include this only to note that a future model could be that projects mutually grant each other tokens in early distributions, to reward actors in the space, without the accusations of self-dealing).

3. A development reward, the unfortunately-named Founder's Reward in Zcash, in which a portion of each block reward includes a distribution to the founders and developers. This distribution can continue for a period of time or be ongoing, and could be allocated directly to a foundation or organization (3a), or to a decentralized app that could then allocate funding directly to developers through a voting mechanism (3b).

## Evaluating Funding Models ##

The desiderata for a funding model for developers are that funding should be sustainable, meaning that under no circumstance should usage of the protocol be high but development underfunded, the model shouldn't unfairly burden one group of stakeholders over another, the model shouldn't be predicated on an unfair distribution of tokens, and it should be difficult to manipulate or attack.

Under these conditions, I would argue that (1) doesn't satisfy the sustainability condition, as it depends on either the generosity of early miners and holders of the token, or the generosity of open source developers to work without compensation. As an ideological point, I believe that it's a general flaw in our software funding model for commercial entities to directly or indirectly profit from the unpaid labor of open source developers that build and maintain projects, and, likewise, users and validators in a decentralized protocol shouldn't rely on the unpaid labor of developers. Moreover, there is a clear record of the dangers of this approach, both in OpenSSL vulnerabilities and more recent [attacks](https://cointelegraph.com/news/bitpays-copay-wallet-compromised-by-malicious-code-firm-issues-advice-for-users).

The counterargument to an Ethereum-style distribution, as in (2), is that allocating a large supply of tokens in a pre-mine (72% in the Ethereum case) prematurely rewards founders and contributors of a protocol without adequately incentivizing them to build the software necessary to run the protocol. Further, a single initial allocation to a foundation presents the possibility that the foundation could liquidate its tokens prematurely, or in a down-cycle, to cover its operating expenses and lose the ability to fund development long-term, failing the sustainability criterion. The fairness arguments against a large pre-mine follow from those that are employed to justify vesting in a typical startup: ownership should be distributed as a function of work, not at the moment of creation, and contributors that leave the project shouldn't receive the same distribution as those that remain.

That leaves the development reward as a viable option for fairly distributing tokens in such a way that funds development long-term. The development reward, without a pre-mine, is more fair, since it doesn't prematurely allocate a large percentage of the supply of the token to insiders. Rather, the token distribution is based on early participation in the protocol, which is open to anyone. The development reward satisfies the sustainability criterion, since as long as the protocol is being used, the token value will not be zero. This model does impose a tax on miners or validators, since part of the block reward is allocated to developers, but this tax can be partially passed on to the user in the form of higher transaction fees, so stakeholders share the cost of funding development. 

The counter-argument to a development reward is that it runs counter to decentralization, in that it provides an ongoing distribution of tokens to a single entity. I would reply that it is less centralized and a more fair distribution than a pre-mine, since it allocates a greater portion of the supply to validators or miners rather than founders, and the development reward could be easily forked out of the protocol (obviously at the expense of halted development), at least compared with rewriting balances to remove non-contributing founders. Further, the initial distribution model of Bitcoin is not viable now, since the entrance of capital and mining equipment in the space, individuals without specialized mining hardware, like Satoshi in 2008, would have no chance to mine blocks unless they did so in secret, which would be an effective pre-mine. Also, relying on the generosity of either holders or developers is not a sustainable or effective solution longterm, whereas Zcash has shown the ability to execute on an ambitious technical roadmap, demonstrating the effectiveness of a development reward. 

The question then is whether the development reward should be distributed to a company or a foundation, or whether it should remain in-protocol and be allocated in a decentralized way, through voting. I think that, while the latter is appealing, voting schemes are too easy to manipulate, and schemes of increased complexity could be hacked or attacked, as famous examples have shown. Therefore, I think that the best model for the space is the Zcash approach, in which a development reward is allocated to an organization. This organization would be "decentralized-through-deterrence": it would be strongly incentivized to act inclusively and transparently, to fairly allocate funding throughout the community and larger open source ecosystem, and to follow the general will of stakeholders in the protocol, since the threat of forking would be its Sword of Damocles, so to speak.

**Comment by dlubarov** [comment]
Great writeup Brendan!

> Under these conditions, I would argue that (1) doesn't satisfy the sustainability condition, as it depends on either the generosity of early miners and holders of the token, or the generosity of open source developers to work without compensation.

Agreed. I think (1) can work for simpler projects, but since our goals are very ambitious, we'll need a handful of researchers and engineers who are really focused on the project for several years. Besides recruiting, it would also be good to have funding for security audits, events, exchange listing fees, etc.

> The counterargument to an Ethereum-style distribution, as in (2), is that allocating a large supply of tokens in a pre-mine (72% in the Ethereum case) prematurely rewards founders and contributors of a protocol without adequately incentivizing them to build the software necessary to run the protocol.

I think under (3), a founder could theoretically still "cash out" early by selling futures on a derivatives market (corresponding to their future vesting events). Maybe it's unlikely to happen in practice though? There might not be a derivatives markets for our token, or liquidity might be low, or founders might just not want to bother with derivatives.

I guess we can view (3) as making it more inconvenient, and potentially costly, for founders to cash out early. Seems like a step in the right direction at least.

> (3) model does impose a tax on miners or validators

Maybe we could implement the development reward by creating a bunch of token records in the genesis, with various time locks for different vesting dates. That would essentially have the same effect, but it might make the protocol a tad simpler. And it might help people understand that the development reward doesn't disproportionally affect miners.

> the development reward could be easily forked out of the protocol (obviously at the expense of halted development)

I guess that's a minor point against I said above. Using time-locked records in the genesis block may make it a little harder to remove the development reward; we'd need to make their nullifiers public.

> Zcash has shown the ability to execute on an ambitious technical roadmap, demonstrating the effectiveness of a development reward.

I mostly agree, but although they've been doing good work, I worry that Zcash's FW will run out before they get to other import projects like scaling. There might be some steps we can take to minimize that risk, like a slower vesting schedule, but I think it mainly comes down to the value of the token. E.g. the Ethereum Foundation still has plenty of funds, since ETH has a much larger market cap than ZEC.

**Comment by bpfarmer** [comment]
> I think under (3), a founder could theoretically still "cash out" early by selling futures on a derivatives market (corresponding to their future vesting events). Maybe it's unlikely to happen in practice though? There might not be a derivatives markets for our token, or liquidity might be low, or founders might just not want to bother with derivatives.

Yeah I think that's a fair point, but, as you note, I don't think that there would be much of a market for derivatives. Plus, I think that the controlling organization, Mir-Corp or Mir-Org, would receive all of the tokens and then pay employees in fiat or grant tokens, in which case I think that it would be more difficult to construct a derivatives market, or at least the rates would be lower.

> I guess that's a minor point against I said above. Using time-locked records in the genesis block would make it nontrivial to remove the development reward, though it can be done of course.

I think that this is worth considering further - whether or not the development reward should be distributed as part of the block reward or whether it should be in the form of time-locked tokens. Both could be forked out of the protocol, but it might be more difficult to generate consensus for a fork that actually removed tokens from circulation, as opposed to a fork that changed a flag to remove the dev reward. I think that the benefit of the time-locked tokens is that it's harder to create frivolous forks, like those occurring in Bitcoin, but it also seems more like a pre-mine. One potential argument for a block reward distribution is that we could lobby the community to fork to allow a greater or lesser development reward rate, depending on cash flow requirements for the org, whereas that would be difficult and awkward with existing tokens.

> E.g. the Ethereum Foundation still has plenty of funds, since ETH has a much larger market cap than ZEC.  

Do they? I was under the impression that they had been forced to sell a lot of ETH at a lower price and as a result were somewhat under-resourced, but I could be totally wrong. I think that removing the vesting schedule and making the dev reward an ongoing distribution might help stabilize funding a little bit, especially if we don't have a capped supply, but a constant inflation rate due to validator rewards. Regardless, the token value will need to be non-zero :)

**Comment by bpfarmer** [comment]
> So your original (3a) proposal sounds good to me.

Yeah I think that it's the best option, but it will be interesting to hear more counterarguments.

> Were you thinking that the development reward would be a fraction of the full mining reward (including fees), or of the block reward (excluding fees)? I'd prefer the latter -- I think it's best for miners to get the full transaction fee, partly to make sure we don't have any financial incentive to influence fees, and partly to make sure users don't have any reason to pay miners out-of-band.

Definitely the latter, I agree with you on fees.

**Comment by bpfarmer** [comment]
Leaving [this](https://medium.com/@tcrypt/instant-sync-full-validation-bitcoin-nodes-9d1b33154a48) here as an illustrative example of why sustainable development funding is necessary. Quick validation through ZKP's is indisputably a good feature for any cryptocurrency, and doesn't even require buy-in at the protocol level, but it doesn't exist in Bitcoin due to a lack of infrastructure funding. Instead, companies like O(1) Labs start, develop what is essentially a single novel feature, and launch their own coins. I suppose that this is unavoidable in an immature technology, but a good argument for why generously funding development is so important.

**Comment by bpfarmer** [comment]
Another [perspective](https://twitter.com/nwoodfine/status/1066941125557026817) :)

**Comment by dlubarov** [comment]
> it doesn't exist in Bitcoin due to a lack of infrastructure funding

Agreed! Bitcoin pretty much seems stuck with its original feature set.

> Another [perspective](https://twitter.com/nwoodfine/status/1066941125557026817) :)

It's surprising to me that criticisms like that tend to focus on Ethereum and Zcash, where founders and investors receive ~20%. To me, distributions like that of Ripple and EOS seem much more "greedy", since all the coins go to a for-profit company. In those cases we don't know how many of the coins will go toward development, and how many will be paid out to shareholders.

I think the best we can do is commit that a certain percentage of coins will go toward development, perhaps by putting them in a non-profit. And let's be very transparent about the allocation. (In the case of Cardano, some coins will go to a non-profit, but I don't think the ever disclosed the amount.)

**Comment by bpfarmer** [comment]
> It's surprising to me that criticisms like that tend to focus on Ethereum and Zcash, where founders and investors receive ~20%. To me, distributions like that of Ripple and EOS seem much more "greedy", since all the coins go to a for-profit company. In those cases we don't know how many of the coins will go toward development, and how many will be paid out to shareholders.

I agree in principle - I think that Zcash should have allocated all coins to the corporation and foundation, instead of distributing to individuals. I think that this is important, because it's difficult to predict the price of coins, so it might be preferable to denominate compensation for founders and employees in dollars rather than in coins, so that you don't end up with a situation where founders are overcompensated because coins outperform expectations in a speculative bubble. Equity holders in the company would still share in the upside, as the company itself would become more valuable as its revenue from the development reward grew. 

For Ethereum, I'm a little bit more skeptical. Assuming that I'm remembering the details correctly, I think that they premined 72% of the total supply of the coin, allocating 12% to founders upfront, and selling the remaining 60%. I think that this is suboptimal for a few reasons: first, there's no vesting element for founders, which is relevant, since most of the founders stopped working directly on Ethereum, and second, I think that selling 60% of the total supply at the price of the token at founding is a bit like selling 60% of a company at the seed stage - it's too much of a discount on the expected future price, and I think that it might hurt other stakeholders early on in the protocol's development by increasing the available supply of tokens. Further, allocating a large portion of the supply of coins directly to founders doesn't really help fund development - there's only a weak incentive to liquidate personal holdings to fund developers - and it just seems patently unfair, as in the Tezos case, for instance, to give personal bonuses of $20m to founders before a testnet is even launched. 

> To me, distributions like that of Ripple and EOS seem much more "greedy", since all the coins go to a for-profit company. In those cases we don't know how many of the coins will go toward development, and how many will be paid out to shareholders.

I definitely agree with that, though I don't think that critics would compare Ripple/EOS favorably with Zcash, instead, I think that Bitcoin's early distribution is the counterargument. As we've discussed, I don't think that Bitcoin's distribution mechanism would be as fair or as effective for funding development now, given the increased interest in the space and proliferation of professional mining operations.

> I think the best we can do is commit that a certain percentage of coins will go toward development, perhaps by putting them in a non-profit. And let's be very transparent about the allocation. (In the case of Cardano, some coins will go to a non-profit, but I don't think the ever disclosed the amount.)

I agree. I think that we should set up a for-profit and divert 100% of the development reward to that organization, which would hire a permanent staff of engineers and researchers to develop and maintain the base protocol, and then, assuming that the development reward grows in dollar terms, establish a foundation that would allocate grants and research funding to other parties, and allocate a percentage of the development reward to that. I agree that we should transparently reveal how the development reward is allocated toward actually developing the protocol.

**Comment by dlubarov** [comment]
> there's no vesting element for founders, which is relevant, since most of the founders stopped working directly on Ethereum

Good point -- that seems like a good reason to have any development rewards go through an organization rather than going directly to individuals.

> I think that selling 60% of the total supply at the price of the token at founding is a bit like selling 60% of a company at the seed stage

Agreed! I think if we did a single ICO (rather than spreading it out like EOS), it should probably be for a smaller quantity of tokens, especially if we're still in early stages.

> it just seems patently unfair, as in the Tezos case, for instance, to give personal bonuses of $20m to founders before a testnet is even launched.

Wow, I hadn't heard about that. Apparently they'll also receive 10% of the token supply over 4 years? (Again I'm having trouble finding any official info.) It's good that it vests, but such a large amount does seem unfair.

> I think that we should set up a for-profit and divert 100% of the development reward to that organization, which would hire a permanent staff of engineers and researchers to develop and maintain the base protocol, and then, assuming that the development reward grows in dollar terms, establish a foundation that would allocate grants and research funding to other parties, and allocate a percentage of the development reward to that.

Similar to what Zcash did? I think that approach is reasonable, but it might be best to allocate funds for the non-profit upfront (even if it won't exist until later), since it benefits all users. I worry that the Zcash Foundation may not last very long, since it will only receive 3% of ZEC (as per Zooko's [zcon0 slides](https://twitter.com/bcrypt/status/1011602573873336320)), which I think came out of the company's funds. I think they could have justified a larger FW at the beginning if everyone knew that a certain amount would go to a non-profit, and that seems better for everyone IMO.

**Comment by bpfarmer** [comment]
> Agreed! I think if we did a single ICO (rather than spreading it out like EOS), it should probably be for a smaller quantity of tokens, especially if we're still in early stages.

I agree. It will be interesting to compare the various initial fundraising options: selling tokens to institution investors, selling equity to institutional investors, or selling tokens to anyone. I think that the last one is clearly unfeasible for regulatory reasons, but among the first two, it will be interesting to see which makes more sense in "crypto winter" conditions. Selling tokens directly has the benefit of providing a means of initial distribution, but it's unclear that investors will actively participate in staking, and it's unclear that such a distribution would be better than distributing via a storage node reward, or selling tokens at a slower pace from the validator reward. I think that it's intellectually interesting to think about how one might avoid a pre-mine entirely, but it's unclear whether we can arrive at a better mode of distribution.

> Similar to what Zcash did? I think that approach is reasonable, but it might be best to allocate funds for the non-profit upfront (even if it won't exist until later), since it benefits all users. I worry that the Zcash Foundation may not last very long, since it will only receive 3% of ZEC (as per Zooko's zcon0 slides), which I think came out of the company's funds. I think they could have justified a larger FW at the beginning if everyone knew that a certain amount would go to a non-profit, and that seems better for everyone IMO.

I think that's a good argument.

**Comment by bpfarmer** [comment]
I've been thinking about this a little bit more lately - I think that if Mir ever became core infrastructure in the Internet, the development reward would be a really good way to fund the "digital commons" consisting of open source projects that are widely used, especially by Mir, but struggle with funding (in addition to funding Mir-specific work). Candidates could include Tor, Rust, academic research, etc. Sort of far down the line, but open source funding is an important and really unsolved problem, and something like the development reward could be a candidate solution.

**Comment by dlubarov** [comment]
That sounds great to me, as long as we can figure out a way to fund it without breaking any expectations. I.e. if it came out of Mir's non-profit's funds, users might expect those funds to be used for Mir development, though I guess it all depends on how we set expectations.

Were you thinking of a permanent development reward? I'd be kinda nervous about that, since (as far as I know?) it hasn't been tried. I think it could be controversial either for financial reasons (if users want minimal inflation), or from concerns about future politics/corruption.

**Comment by bpfarmer** [comment]
> Were you thinking of a permanent development reward? I'd be kinda nervous about that, since (as far as I know?) it hasn't been tried. I think it could be controversial either for financial reasons (if users want minimal inflation), or from concerns about future politics/corruption.

Yes, and I think that it would be fairly controversial for sure! It hasn't been attempted yet, but it's fairly likely that ZEC will have to fork to extend the reward, since otherwise they'll have a funding gap.

That said, a distribution structured like the following: no premine at all, and then a block reward that is split between the block producer, the Mir company, and the Mir foundation (as well as potentially storage nodes?), with the block producer receiving the majority of the reward, is definitely more fair (in my opinion) that the Ethereum distribution (or every ICO), and arguably more fair than the Bitcoin distribution. I'm not sure whether this is feasible with Proof-of-Stake, or whether you'd need to bootstrap a wider distribution of stakers initially (perhaps participants could run storage nodes to accumulate enough tokens for a validator slot).

If you assume perpetual inflation and the BP reward of 2/3 of the block reward, while the remaining 1/3 is distributed jointly between the company and the foundation, then 2/3 of the total supply of tokens at any time will have been distributed to block producers, and 1/3 to Mir [Corp/Org]. This seems like a lot, except when you compare to Ethereum, which gave founders 12% of the total supply up front, and then sold 60% of the remaining supply up-front (effectively giving it to the foundation and then selling it, not to mention the self-dealing that occurred when founders with sufficient capital bought tokens at a lower price). Therefore, when the supply cap is reached (though presumably inflation will continue), only 28% of the total supply of tokens will have been distributed to block producers, and moreover, 72% of the total supply was given to founders and the foundation before even a testnet, whereas Mir will have to constantly justify the Development Reward or risk a fork with no Development Reward becoming dominant.

Bitcoin's distribution mechanism isn't feasible, as investment in mining has put it outside the reach of the casual hobbyists that became early bitcoin miners, and it wouldn't be possible for the developers of a protocol to mine enough coins to fund operations. Even so, I'd argue that Bitcoin was unfairly distributed, as much of the token supply was distributed on the basis of luck - to hobbyists that happened to be capable of setting up a mining node and were exposed to Bitcoin early enough - rather than desert, which to me, would've involved a greater distribution to the writers of the software and active participants in the network.  

The trick with a perpetual development reward is guaranteeing that it's distributed to the most effective development teams, and there's no guarantee that this will always be Mir [Corp/Org]. However, I think that competing proposals like DAO's are unworkable due to complexity and coordinated attacks, so it seems like a development reward directed to an off-chain institution is the current best option, even at the expense of centralization. Plus, there will be a strong incentive for the Org to support the best teams, and for the Corp to hire the best cryptographers, researchers, and engineers, as development failure will lead to adoption of another protocol or of a fork that doesn't include the development reward (or reduces it).
 
I think that there's also the pragmatic case: a protocol like Mir will require iterative development beyond the first mainnet release, and it seems impossible or difficult to depend on the open source community for that development (not to mention, open source developers shouldn't be essentially subsidizing stakers and early holders). ZEC has managed to produce two technically complex but successful releases in a fairly short time, whereas other protocols like Ethereum have struggled to coordinate a semi-centralized collection of underfunded teams.  

There probably should be another issue or blog post on decentralization, in particular that centralizing the distribution of development funds via a Corp/Org doesn't break other desirable properties of decentralized protocols. But, this is again fairly controversial, and we should continue the conversation!

**Comment by dlubarov** [comment]
> Mir will have to constantly justify the Development Reward or risk a fork with no Development Reward becoming dominant.

That's a great point. As a user, I'd definitely prefer a dev reward that vests slowly over an immediate reward, for that reason.

I still worry a bit about an indefinite reward though, just because the riskiness of forks means that defaults tend to stick. Like in 50 years if the dev reward is no longer being used effectively, the community might reason that the cost of the dev reward is small compared to the possible cost of splitting the network with a controversial fork.

I like the idea of a large reward which vests very slowly, like over 20 years or something. We could consider proposing a hard fork after that, if we felt more funding was necessary, but the default rules would be against that, to guard against the above scenario.
